updatehub-sdk/src/listener.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2018, 2019, 2020 O.S. Systems Sofware LTDA
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Error, Result};
+use async_ctrlc::CtrlC;
+use async_std::{
+    io::BufReader,
+    os::unix::net::{UnixListener, UnixStream},
+    prelude::*,
+};
+use slog::*;
+use std::{collections::HashMap, net::Shutdown, process};
+
+const SDK_TRIGGER_FILENAME: &str =
+    "/usr/share/updatehub/state-change-callbacks.d/10-updatehub-sdk-statechange-trigger";
+const SOCKET_PATH: &str = "/run/updatehub-statechange.sock";
+
+type StateChangeCallback = Box<dyn Fn(State)>;
+type ErrorCallback = Box<dyn Fn(String)>;
+
+#[derive(Default)]
+pub struct StateChangeListener {
+    listeners: HashMap<StateID, Vec<StateChangeCallback>>,
+    error_handlers: Vec<ErrorCallback>,
+}
+
+#[derive(PartialEq, Eq, Hash)]
+pub enum StateID {
+    Download,
+    Install,
+    Reboot,
+    Error,
+}
+
+pub struct State {
+    pub id: StateID,
+    pub conn: UnixStream,
+}
+
+impl State {
+    pub fn cancel(&mut self) {
+        self.conn.write_all(b"cancel");
+    }
+
+    pub fn proceed(&self) {}
+}
+
+impl StateChangeListener {
+    pub fn new() -> Self {
+        StateChangeListener { listeners: HashMap::new(), error_handlers: Vec::new() }
+    }
+
+    pub fn on(&mut self, state: StateID, f: StateChangeCallback) {
+        match self.listeners.get_mut(&state) {
+            Some(callbacks) => {
+                callbacks.push(Box::new(f));
+            }
+            None => {
+                self.listeners.insert(state, vec![Box::new(f)]);
+            }
+        }
```

Comment:
Take a look at the `Entry`[1] API. I think this could be of use here to simplify the code.

1. https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html

Reply:
Have you checked the `entry` API? it was not changed.

Reply:
I checked and tried, however, as I understood it would only work for the case where there is no value associated with the callback vector.

Reply:
I think you missed this: https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.and_modify

Reply:
What if we didn't use a HashMap at all? It doesn't seem quite like the right data structure for this. We have a predefined and limited number of possible keys with a unlimited number of values for each key. I think that even a static list of vectors would be more suitable; either by using a slice or using named fields since we only have four states to handle. This would allow us to avoid this costs (both on code complexity and on execution) of having to check if a list is present before appending to it.

Reply:
I agree with your point of HashMap is a little exaggerated to solve the problem, since we know exactly the number of states (on context of HashMap, the keys).